pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+                return index;
+        }
+        else if (!(memcmp(item, &elements[index * tSizes[list->type]], tSizes[list->type])))
+                return index;
+        index += 1;
+    }
+    return (-1);
+}
+
+size_t calculate_index(long int index, size_t size)
+/*
+        calculates negative indexes
+
+        Parameters
+        ----------
+        index : long int
+            the index to be resolved
+
+        size : size_t
+            the element to be indexed
+
+        Returns
+        =======
+        index : size_t
+*/
+{
+    if (index >= size)
+        index = size;
+    else if (index < 0)
+    {
+        index += size + 1;
+        index = (index > 0) ? index : 0;
+    }
+    return index;
+}
+
+void     insert(t_list* list, long int index, void* item)
+/*
+        Insert an element in a specific index
+
+        Parameters
+        ----------
+        list : t_list*
+            t_list where to insert the element
+
+        index : long int
+            the index where to insert the element
+
+        item : void*
+            the element to be inserted
+*/
+{
+    char * elements = list->elements;
+    size_t totalSize = list->size + 1;
+    size_t tsize = tSizes[list->type];
+    size_t ind = calculate_index(index, list->size) * tsize;
+
+    if (totalSize >= list->capacity)
+    {
+        list->capacity *= 2;
+        elements = realloc(elements, list->capacity * tsize);
+        list->elements = elements;
+    }
+    memmove(&(elements[ind + tsize]), &(elements[ind]), (list->size * tsize) - ind);
+    memcpy(&(elements[ind]), item, tsize);
+    list->size = totalSize; 
+}
+
+t_pop_ret   *pop(t_list* list, long int index)
+/*
+        removes an element from a specific index and return it
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list from where to remove the element
+
+        index : long int
+            index of the element to be removed
+
+        Returns
+        =======
+        ret_val : t_pop_ret
+*/
+{
+    size_t tsize = tSizes[list->type];
+    char *elements = (char*)(list->elements);
+    t_pop_ret *ret_val = malloc(sizeof(t_pop_ret));
+
+    index = (index < 0) ? list->size + index : index;
+    if (index >= 0 && list->size > index)
+    {
+        ret_val->type = list->type;
+        if (list->type == lst_list)
+            ret_val->raw = (char *)((t_list**)elements)[index];
+        else
+        {
+            ret_val->raw = malloc(list->type);
+            memcpy(ret_val->raw, &elements[index * tsize], tsize);
+        }
+        memmove(&elements[index * tsize], &elements[(index + 1) * tsize], list->size - index);
+        list->size -= 1;
+        return (ret_val);
+    }
+
+    return (NULL);
+}
+
+void    free_pop(t_pop_ret** pop_val)
+/*
+        Deallocate the t_pop_ret object resulting from a pop operation
+
+        Parameters
+        ----------
+        pop_val : t_pop_ret**
+            double pointer on the t_pop_ret object to be deallocated
```

Comment:
Why double pointer?

Reply:
just so we can put NULL to the pointer we are holding, we could potentially use it to make checks.